### DISCUSSION TITLE
Running out of disk space, how to troubleshoot?

### DISCUSSION THREAD

**Discussion by langri-sha** [community/community#46496] [Codespaces] [answered]
Hi everyone! I changed my codespace's machine type to 8-core, to get 64G of disk space. I'm constantly running out of space, and this is what I get with `ncdu /`:

![image](https://user-images.githubusercontent.com/77084/216966388-c339f95e-3875-452c-a039-586d8545f56f.png)

So it looks as though I should have plenty to spare, but then looking at `df -h`:

```
Filesystem      Size  Used Avail Use% Mounted on
overlay          63G   57G  3.2G  95% /
tmpfs            64M     0   64M   0% /dev
tmpfs           7.9G     0  7.9G   0% /sys/fs/cgroup
shm              64M     0   64M   0% /dev/shm
/dev/sda1        29G   15G   15G  51% /usr/sbin/docker-init
/dev/loop0       63G   57G  3.2G  95% /workspaces
/dev/sdb1        63G  192K   60G   1% /tmp
```

Any suggestions on what I could do? Last time I've reached this point, I destroyed the codespace and started over, but I'd like to avoid doing that this time around.

**Comment by langri-sha** [comment, marked as answer]
After running `sudo ncdu /` instead I was able to notice a lot of space taken up by dangling Docker build caches 🌞.

**Reply by nnWhisperer** [reply]
Hi! How were you able to reclaim this space? I recently did a full rebuild of my Codespace, but it didn't free up the 12GB after purging docker build caches. Also, I saw a comment suggesting to delete the /workspaces/.codespaces/shared/editors/jetbrains folder, but that directory isn't accessible or removable on my end.

**Reply by jeson114** [reply]
docker builder prune

**Reply by nnWhisperer** [reply]
> docker builder prune

@jeson114 
I'm talking about the 12GB disk space that was already consumed at the start of codespace and docker builder prune (--all) didn't release it.

**Reply by KaiquanMah** [reply]
@nnWhisperer 

These commands below freed up very little space on github codespace for me:
```bash
sudo apt-get clean
pip cache purge
docker system prune -f
docker volume prune -f


# check diskspace
df -h
Filesystem      Size  Used Avail Use% Mounted on
overlay          32G   29G  1.7G  95% /
tmpfs            64M     0   64M   0% /dev
shm              64M     0   64M   0% /dev/shm
/dev/root        29G   22G  7.6G  74% /vscode
/dev/sdc1        44G  3.9G   38G  10% /tmp
/dev/loop4       32G   29G  1.7G  95% /workspaces
```


I also seemed to be unable to access the recycle bin, where I feel there might be things clogging up the storage in my codespace.
```bash
ls -la ~/.local/share/Trash/*
ls: cannot access '/home/codespace/.local/share/Trash/*': No such file or directory
```


As a last resort, **I went to the top of codespace >> `> rebuild codespace` >> full rebuild.**
This tip came from: https://www.reddit.com/r/cs50/comments/u1ywb0/no_space_left_on_device_codespaces.

```bash
# check diskspace

df -h
Filesystem      Size  Used Avail Use% Mounted on
overlay          32G   12G   18G  40% /
tmpfs            64M     0   64M   0% /dev
shm              64M     0   64M   0% /dev/shm
/dev/root        29G   22G  7.6G  74% /vscode
/dev/sdc1        44G  4.2G   38G  10% /tmp
/dev/loop4       32G   12G   18G  40% /workspaces
```

**Comment by Ronnie-personal** [comment]
@langri-sha  Would you mind if I ask a question? How did you get 64GB space on the '/' volume? I tried to create a new codespace or change machine type for an existing codespace, however the only option for disk is 32GB.

**Reply by AndnixSH** [reply]
I believe he have access to better machines with 64 gb space. I still don't have access to better machines even I subscribed to Github PRO

**Reply by Ronnie-personal** [reply]
Same here.
I upgraded my personal GitHub account from the free to the PRO plan; however, the available machine types for Codespaces remain the same, with options for 2 or 4 cores and 32GB of disk space.

**Comment by MikeRecognex** [comment]
My workspace had jetbrains installed by default and this was taking 12 GB.   Is this normal..?  Also please see below issue with docker

@xxxxx ➜ /workspaces/xx (main) $ du -ah /workspaces | sort -rh | head -20
12G     /workspaces/.codespaces/shared/editors/jetbrains
12G     /workspaces/.codespaces/shared/editors
12G     /workspaces/.codespaces/shared
12G     /workspaces/.codespaces
12G     /workspaces
3.6G    /workspaces/.codespaces/shared/editors/jetbrains/idea-IU-231.8770.65
3.4G    /workspaces/.codespaces/shared/editors/jetbrains/JetBrainsRider-2023.1
2.4G    /workspaces/.codespaces/shared/editors/jetbrains/idea-IU-231.8770.65/plugins
2.3G    /workspaces/.codespaces/shared/editors/jetbrains/pycharm-2023.1.1
2.1G    /workspaces/.codespaces/shared/editors/jetbrains/RubyMine-2023.1.1
1.4G    /workspaces/.codespaces/shared/editors/jetbrains/JetBrainsRider-2023.1/lib
1.3G    /workspaces/.codespaces/shared/editors/jetbrains/JetBrainsRider-2023.1/plugins
1.1G    /workspaces/.codespaces/shared/editors/jetbrains/pycharm-2023.1.1/plugins
886M    /workspaces/.codespaces/shared/editors/jetbrains/RubyMine-2023.1.1/plugins
665M    /workspaces/.codespaces/shared/editors/jetbrains/idea-IU-231.8770.65/lib
631M    /workspaces/.codespaces/shared/editors/jetbrains/pycharm-2023.1.1/lib
605M    /workspaces/.codespaces/shared/editors/jetbrains/RubyMine-2023.1.1/lib
601M    /workspaces/.codespaces/shared/editors/jetbrains/JetBrainsRider-2023.1/lib/app.jar
596M    /workspaces/.codespaces/shared/editors/jetbrains/JetBrainsRider-2023.1/lib/ReSharperHost
502M    /workspaces/.codespaces/shared/editors/jetbrains/idea-IU-231.8770.65/lib/app.jar
@xx ➜ /workspaces/xx (main) $ find ~ -name '*jetbrains*' -exec rm -rf {} +

@xx➜ /workspaces/xx (main) $ 
@xx ➜ /workspaces/xx (main) $ du -ah /workspaces | sort -rh | head -20
12G     /workspaces/.codespaces/shared/editors/jetbrains
12G     /workspaces/.codespaces/shared/editors
12G     /workspaces/.codespaces/shared
12G     /workspaces/.codespaces
12G     /workspaces
3.6G    /workspaces/.codespaces/shared/editors/jetbrains/idea-IU-231.8770.65
3.4G    /workspaces/.codespaces/shared/editors/jetbrains/JetBrainsRider-2023.1
2.4G    /workspaces/.codespaces/shared/editors/jetbrains/idea-IU-231.8770.65/plugins
2.3G    /workspaces/.codespaces/shared/editors/jetbrains/pycharm-2023.1.1
2.1G    /workspaces/.codespaces/shared/editors/jetbrains/RubyMine-2023.1.1
1.4G    /workspaces/.codespaces/shared/editors/jetbrains/JetBrainsRider-2023.1/lib
1.3G    /workspaces/.codespaces/shared/editors/jetbrains/JetBrainsRider-2023.1/plugins
1.1G    /workspaces/.codespaces/shared/editors/jetbrains/pycharm-2023.1.1/plugins
886M    /workspaces/.codespaces/shared/editors/jetbrains/RubyMine-2023.1.1/plugins
665M    /workspaces/.codespaces/shared/editors/jetbrains/idea-IU-231.8770.65/lib
631M    /workspaces/.codespaces/shared/editors/jetbrains/pycharm-2023.1.1/lib
605M    /workspaces/.codespaces/shared/editors/jetbrains/RubyMine-2023.1.1/lib
601M    /workspaces/.codespaces/shared/editors/jetbrains/JetBrainsRider-2023.1/lib/app.jar
596M    /workspaces/.codespaces/shared/editors/jetbrains/JetBrainsRider-2023.1/lib/ReSharperHost
502M    /workspaces/.codespaces/shared/editors/jetbrains/idea-IU-231.8770.65/lib/app.jar
@xxxxxx ➜ /workspaces/xx (main) $ rm -rf /workspaces/.codespaces/shared/editors/jetbrains
@xxxxxxx ➜ /workspaces/xx (main) $ du -ah /workspaces | sort -rh | head -20
24M     /workspaces
22M     /workspaces/x
21M     /workspaces/xx/src/haystack-ai
21M     /workspaces/xx/src
11M     /workspaces/xxsrc/haystack-ai/.git/objects/pack
11M     /workspaces/xx/src/haystack-ai/.git/objects
11M     /workspaces/xx/src/haystack-ai/.git
5.3M    /workspaces/xx/src/haystack-ai/.git/objects/pack/pack
4.3M    /workspaces/xx/src/haystack-ai/.git/objects/pack/.pack
4.0M    /workspaces/xx/src/haystack-ai/docs
3.9M    /workspaces/xx/src/haystack-ai/docs/img
2.2M    /workspaces/xx/src/haystack-ai/test
1.7M    /workspaces/.codespaces
1.4M    /workspaces/xx/src/haystack-ai/releasenotes/notes
1.4M    /workspaces/xx/src/haystack-ai/releasenotes
1.4M    /workspaces/xx/src/haystack-ai/haystack
1.3M    /workspaces/.codespaces/.persistedshare
1.1M    /workspaces/xx/src/haystack-ai/test/test_files
1.1M    /workspaces/xx/src/haystack-ai/docs/img/main_example.gif
852K    /workspaces/xx/src/haystack-ai/proposals
@xxxxxxxx ➜ /workspaces/Recostack (main) $ du -ah /workspaces | sort -rh | head -20
24M     /workspaces
22M     /workspaces/xx
21M     /workspaces/xx/src/haystack-ai
21M     /workspaces/xx/src
11M     /workspaces/xx/src/haystack-ai/.git/objects/pack
11M     /workspaces/xx/src/haystack-ai/.git/objects
11M     /workspaces/xx/src/haystack-ai/.git
5.3M    /workspaces/xx/src/haystack-ai/.git/objects/pack/pack-bad9dd3fd5b3d8ac89c7bbeb03b40158bdf1ac64.pack
4.3M    /workspacesxx/src/haystack-ai/.git/objects/pack/pack-85567a276ab53de73f88e71a997f85884385ef2d.pack
4.0M    /workspaces/xxx/src/haystack-ai/docs
3.9M    /workspaces/xxsrc/haystack-ai/docs/img
2.2M    /workspaces/xx/src/haystack-ai/test
1.7M    /workspaces/.codespaces
1.4M    /workspaces/xx/src/haystack-ai/releasenotes/notes
1.4M    /workspaces/xx/src/haystack-ai/releasenotes
1.4M    /workspaces/xxsrc/haystack-ai/haystack
1.3M    /workspaces/.codespaces/.persistedshare
1.1M    /workspaces/xxsrc/haystack-ai/test/test_files
1.1M    /workspaces/xx/src/haystack-ai/docs/img/main_example.gif
852K    /workspaces/xx/src/haystack-ai/proposals


please explain below....

@xxx ➜ /workspacesxxx(main) $ df -h /var/lib/docker
Filesystem      Size  Used Avail Use% Mounted on
/dev/loop3       32G   15G   16G  49% /var/lib/docker
@xx➜ /workspaces/xx(main) $ cd /var/lib/docker
bash: cd: /var/lib/docker: Permission denied
@xxx ➜ /workspaces/xx (main) $ ls /var/lib/docker
ls: cannot open directory '/var/lib/docker': Permission denied
@xxx➜ /workspaces/xx (main) $ docker images --format "table {{.Repository}}\t{{.Tag}}\t{{.Size}}"
REPOSITORY   TAG       SIZE
@xx ➜ /workspaces/xxx (main) $ docker images -f "dangling=true" -q
@xx➜ /workspaces/xx (main) $ docker builder prune
WARNING! This will remove all dangling build cache. Are you sure you want to continue? [y/N] y
Total:  0B
@xx➜ /workspaces/xxx(main) $ docker volume ls
DRIVER    VOLUME NAME
@xxx ➜ /workspaces/xxx(main) $ docker system prune -a --volumes
WARNING! This will remove:
  - all stopped containers
  - all networks not used by at least one container
  - all anonymous volumes not used by at least one container
  - all images without at least one container associated to them
  - all build cache

Are you sure you want to continue? [y/N] y
Total reclaimed space: 0B
@xx➜ /workspaces/Recostack (main) $ df -h /var/lib/docker
Filesystem      Size  Used Avail Use% Mounted on
/dev/loop3       32G   15G   16G  49% /var/lib/docker
@xx ➜ /workspaces/xx (main) $

**Comment by o8e** [comment]
`rm -rf /workspaces/.codespaces/shared/editors/jetbrains`

If you don't need any preinstalled editors. That'll give you ~11gb back.

**Reply by nnWhisperer** [reply]
I could neither access nor remove it. I guess it doesn't work anymore.

**Comment by mohhasbias** [comment]
Just sharing, if you are using Docker, you can check their disk usage using this command `docker system df`

here is an example output
<img width="433" height="73" alt="image" src="https://github.com/user-attachments/assets/2637dc2a-7aaf-4cae-8fdf-fb7c7e175b0e" />

**Comment by mdlsthkarmd54-creator** [comment]
https://ci3.googleusercontent.com/meips/ADKq_NY5JJ-g2zf0qdsGzyewozvQahd5EliXng91jq0wC3iekrNhcKoJYfYIfElHfial43uEoVASU3gW1yyounhHNU7t_RZXDCiELde_q-qsSF3oIXjfMmxq_dYWnz0P=s0-d-e1-ft#https://qph.fs.quoracdn.net/main-qimg-f50624e681101d2b3bc329abc837d7a2

**Comment by ASHLIN-JOHN** [comment]
What you’re seeing is actually normal behavior in GitHub Codespaces, even though it’s confusing.

What’s going on?

/ and /workspaces are backed by an overlay filesystem (mounted via /dev/loopX)

Docker data lives on a separate loop-mounted disk at /var/lib/docker

You cannot access or delete /var/lib/docker directly (Permission denied is expected)

docker system prune, docker builder prune, etc. only clean Docker-managed objects, not space reserved by the Codespace image itself

So when df -h shows something like:

/dev/loop3 32G 15G used


but:

docker images → empty

docker system prune → 0B reclaimed

that means the space is preallocated or reserved by the base Codespace image, not dangling images or volumes.

About JetBrains taking ~12GB

Older Codespaces images used to install JetBrains IDEs under:

/workspaces/.codespaces/shared/editors/jetbrains


That folder used to be removable, but on newer images it’s:

read-only, or

recreated automatically

So deleting it no longer works reliably (or at all)

Why ncdu doesn’t show the usage

ncdu / without sudo cannot see protected mounts

Even with sudo, loop-backed storage and overlay layers can look “empty” while still consuming space

The uncomfortable truth 😅

If:

Docker prune reclaims nothing

You can’t access /var/lib/docker

JetBrains folders can’t be removed

Disk usage stays high

➡️ A full Codespace rebuild is the only way to reclaim that reserved space

This is a known limitation of Codespaces, not something you’re doing wrong.

Tips to reduce hitting this again

Use Dev Containers with minimal base images

Avoid large Docker builds inside Codespaces

Periodically do Full Rebuild instead of normal rebuild

Move large temp files to /tmp (which is on a different mount)

TL;DR
You’re not missing a cleanup command. The “used” space is reserved by the Codespace image and Docker storage layers. When prune commands reclaim 0B, a full rebuild is the only real fix.

**Comment by PallabBG** [comment]
Use ncdu or du -sh * (Linux) / Storage Sense or WizTree (Windows) to identify large files, then clear logs, temp files, or cache to reclaim space.
